### PR TITLE
[SYCL] PI interface.

### PR DESCRIPTION
**Pull request by romanovvlad**
Signed-off-by: Vlad Romanov <vlad.romanov@intel.com>

### CODE REVIEW

**Comment by romanovvlad** [comment]
> 
> 
> The concept is super useful.
> We need this to have SYCL more general!
> But SYCL is about modern C++ and reimplementing a C-like ICD in SYCL seems not like the right abstraction in 2019. OpenCL is written in C with its limitations...
> You are reimplementing dynamic polymorphism in C while it is already available in C++.
> An alternative approach can be to have the SYCL implementation classes inheriting from some more or less abstract classes and the SYCL runtime can just use transparently a launch kernel virtual member function of the device class to redirect to the actual kernel launch.
> The ICD C-like method does not solve well the adaptation problem because it forces at the first place to encapsulate any runtime (even when written in C++!) to the same low-level C method...
> I planned to come with an example of this in triSYCL about using dynamic polymorphism for these extensions [triSYCL/triSYCL#237](https://github.com/triSYCL/triSYCL/issues/237) but we are late... I made some presentation in the SYCL committee with some ideas and in triSYCL there is already the use of dynamic polymorphism but without the ability to inject different runtime on the fly yet...
> Your plug-in mechanism can be a class inheriting from the runtime interface and implementing various operations or customization points.
> A multilayered abstraction seems better suited than going directly down to the metal, because probably some SYCL implementation parts will be common to some classes of devices or API (for example some devices require eager buffer allocation, other lazy late allocation, and this does not fit well is a low-level only approach).

Hi Ronan,

Thank you for you comments!

We had to completely rework the patch, new patch is here #222. Some of your
comments were addressed, some weren't yet. Unfortunately, we can't move the
comments to the new pull request. Most of the design related comments still
applicable to the new patch, so let's discuss them there.

File: sycl/include/CL/sycl/detail/pi.h [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+/// Kinds of device images
+enum pi_device_image_format {
+  PI_IMG_NONE,   // image format is not determined
+  PI_IMG_NATIVE, // image format is specific to a device
+  // portable image kinds go next
+  PI_IMG_SPIRV,         // SPIR-V
+  PI_IMG_LLVMIR_BITCODE // LLVM bitcode
+};
+
+typedef void __tgt_offload_entry;
```
[the remark this answers is not in the document]
Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h]: We do want to keep the core pi.h interfaces to be C-only for greater interoperability with non-C++ environments, which is envisioned to be needed. The integration of pi.h into SYCL is done with C++ and all the good modern stuff. Please review #222 and leave any further comments there.

File: sycl/include/CL/sycl/detail/pi.h [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+// changing function signature.
+#define PI_VERSION 1
+
+/// Target identification strings
+#define PI_TGT_STR_UNKNOWN "<unknown>"
+#define PI_TGT_STR_SPIRV32 "spir"
+#define PI_TGT_STR_SPIRV64 "spir64"
+
+/// Kinds of device images
+enum pi_device_image_format {
```
[the remark this answers is not in the document]
Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h]: Same comment about C-only pi.h

File: sycl/include/CL/sycl/detail/pi.h [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+  // portable image kinds go next
+  PI_IMG_SPIRV,         // SPIR-V
+  PI_IMG_LLVMIR_BITCODE // LLVM bitcode
+};
+
+typedef void __tgt_offload_entry;
+
+// Device image descriptor version supported by this library.
+#define PI_DEVICE_IMAGE_STRUCT_VERSION ((uint16_t)1)
+#define SYCL_OFFLOAD_KIND ((uint8_t)4)
```
[the remark this answers is not in the document]
Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h]: OK, please review #222 and leave any further comments there.

File: sycl/include/CL/sycl/detail/pi.h [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+/// It must match the __tgt_device_image structure generated by
+/// the clang-offload-wrapper tool when their Version field match.
+struct pi_device_image {
+  /// version of this structure - for backward compatibility;
+  /// all modifications which change order/type/offsets of existing fields
+  /// should increment the version.
+  uint16_t Version;
+  /// the kind of offload model the image employs; must be 4 for SYCL
+  uint8_t Kind;
+  /// format of the image data - SPIRV, LLVMIR bitcode,...
```
[the remark this answers is not in the document]
Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h]: Right, this is how C-interface implements it.

[the remark this answers is not in the document]
Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h]: please review #222 and leave any further comments there.

File: sycl/include/CL/sycl/detail/pi.h [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+  const unsigned char *ManifestStart;
+  /// Pointer to the manifest data end
+  const unsigned char *ManifestEnd;
+  /// Pointer to the target code start
+  const unsigned char *ImageStart;
+  /// Pointer to the target code end
+  const unsigned char *ImageEnd;
+  /// the offload entry table (not used, for compatibility with OpenMP)
+  __tgt_offload_entry *EntriesBegin;
+  __tgt_offload_entry *EntriesEnd;
```
[the remark this answers is not in the document]
Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h]: I moved this comment to #222 for any further discussion there.

File: sycl/source/detail/pi.cpp [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+
+#include <assert.h>
+#include <cstdint>
+
+cl_int piSelectDeviceImage(pi_context ctx, pi_device_image **images,
+                           cl_uint num_images,
+                           pi_device_image **selected_image) {
+  // TODO dummy implementation.
+  // Real implementaion will use the same mechanism OpenCL ICD dispatcher
+  // uses. Somthing like:
```
[the remark this answers is not in the document]
Comment (smaslov-intel) [reply sycl/source/detail/pi.cpp]: This comment moved to #222 for any further discussion there.

**Pull request closed without merging**